File: app/src/main/java/com/moufee/purduemenus/ui/menu/MenuRecyclerViewAdapter.java
package com.moufee.purduemenus.ui.menu;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.moufee.purduemenus.databinding.FragmentMenuitemBinding;
import com.moufee.purduemenus.databinding.StationHeaderBinding;
import com.moufee.purduemenus.repository.data.menus.MenuItem;
import com.moufee.purduemenus.repository.data.menus.Station;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Created by Ben on 9/7/17.
 * A RecyclerViewAdapter for Menu Items
 */

public class MenuRecyclerViewAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<Object> mListItems = new ArrayList<>();
    private Set<String> mFavoriteSet = new HashSet<>();
    private OnToggleFavoriteListener mFavoriteListener;

    private final static int VIEW_TYPE_HEADER = 0;
    private final static int VIEW_TYPE_ITEM = 1;

    public MenuRecyclerViewAdapter(OnToggleFavoriteListener favoriteListener) {
        mFavoriteListener = favoriteListener;
    }

    public void setFavoriteSet(Set<String> favoriteSet) {
        mFavoriteSet = favoriteSet;
        notifyDataSetChanged();
    }

    public void setStations(List<Station> stations) {
        mListItems.clear();
        for (Station station : stations) {
            if (station.getItems().size() > 0) {
                mListItems.add(station.getName());
                mListItems.addAll(station.getItems());
            }
        }
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater =
                LayoutInflater.from(parent.getContext());
        if (viewType == VIEW_TYPE_HEADER) {
            StationHeaderBinding binding = StationHeaderBinding.inflate(layoutInflater, parent, false);
            return new StationHeaderViewHolder(binding);
        }
        FragmentMenuitemBinding itemBinding =
                FragmentMenuitemBinding.inflate(layoutInflater, parent, false);
        return new MenuItemHolder(itemBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder.getItemViewType() == VIEW_TYPE_HEADER && holder instanceof StationHeaderViewHolder) {
            StationHeaderViewHolder stationHolder = (StationHeaderViewHolder) holder;
            stationHolder.bind((String) mListItems.get(position));
        } else {
            MenuItemHolder itemHolder = (MenuItemHolder) holder;
            MenuItem item = (MenuItem) mListItems.get(position);
            itemHolder.bind(item, mFavoriteListener, mFavoriteSet.contains(item.getId()));
        }
    }

    @Override
    public int getItemCount() {
        return mListItems.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (mListItems.get(position) instanceof String)
            return VIEW_TYPE_HEADER;
        return VIEW_TYPE_ITEM;
    }

    public boolean isHeader(int position) {
        return getItemViewType(position) == VIEW_TYPE_HEADER;
    }

    private class StationHeaderViewHolder extends RecyclerView.ViewHolder {

        private StationHeaderBinding binding;

        StationHeaderViewHolder(StationHeaderBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        private void bind(String stationName) {
            binding.setStationName(stationName);
            binding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/moufee/purduemenus/ui/menu/MenuRecyclerViewAdapter.kt
package com.moufee.purduemenus.ui.menu

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.moufee.purduemenus.databinding.FragmentMenuitemBinding
import com.moufee.purduemenus.databinding.StationHeaderBinding
import com.moufee.purduemenus.repository.data.menus.MenuItem
import com.moufee.purduemenus.repository.data.menus.Station
import java.util.*

/**
 * Created by Ben on 9/7/17.
 * A RecyclerViewAdapter for Menu Items
 */
class MenuRecyclerViewAdapter(private val mFavoriteListener: OnToggleFavoriteListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private val mListItems: MutableList<Any> = ArrayList()
    private var mFavoriteSet: Set<String> = HashSet()
    fun setFavoriteSet(favoriteSet: Set<String>) {
        mFavoriteSet = favoriteSet
        notifyDataSetChanged()
    }

    fun setStations(stations: List<Station>) {
        mListItems.clear()
        for ((name, items) in stations) {
            if (items.isNotEmpty()) {
                mListItems.add(name)
                mListItems.addAll(items)
            }
        }
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        if (viewType == VIEW_TYPE_HEADER) {
            val binding = StationHeaderBinding.inflate(layoutInflater, parent, false)
            return StationHeaderViewHolder(binding)
        }
        val itemBinding = FragmentMenuitemBinding.inflate(layoutInflater, parent, false)
        return MenuItemHolder(itemBinding)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder.itemViewType == VIEW_TYPE_HEADER && holder is StationHeaderViewHolder) {
            holder.bind(mListItems[position] as String)
        } else {
            val itemHolder = holder as MenuItemHolder
            val item = mListItems[position] as MenuItem
            itemHolder.bind(item, mFavoriteListener, mFavoriteSet.contains(item.id))
        }
    }

    override fun getItemCount(): Int {
        return mListItems.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (mListItems[position] is String) VIEW_TYPE_HEADER else VIEW_TYPE_ITEM
    }

    fun isHeader(position: Int): Boolean {
        return getItemViewType(position) == VIEW_TYPE_HEADER
    }

    private inner class StationHeaderViewHolder(private val binding: StationHeaderBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(stationName: String) {
            binding.stationName = stationName
            binding.executePendingBindings()
        }
    }

    companion object {
        private const val VIEW_TYPE_HEADER = 0
        private const val VIEW_TYPE_ITEM = 1
    }
}